sdk/storage/Azure.Storage.Common/src/Shared/LazyLoadingReadOnlyStream.cs

```diff
@@ -0,0 +1,240 @@
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => ReadInternal(
+                async: false,
+                buffer,
+                offset,
+                count,
+                default)
+            .EnsureCompleted();
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            => await ReadInternal(
+                async: true,
+                buffer,
+                offset,
+                count,
+                cancellationToken)
+                .ConfigureAwait(false);
+
+        public async Task<int> ReadInternal(bool async, byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ValidateReadParameters(buffer, offset, count);
+
+            if (_stream.Position == 0)
+            {
+                await Download(async, cancellationToken).ConfigureAwait(false);
+                if (_lastDownloadBytes == 0)
+                {
+                    return 0;
+                }
+            }
+
+            int totalCopiedBytes = 0;
+            do
+            {
+                int copiedBytes = async
+                    ? await _stream.ReadAsync(buffer, offset, count).ConfigureAwait(false)
+                    : _stream.Read(buffer, offset, count);
+                offset += copiedBytes;
+                count -= copiedBytes;
+                _position += copiedBytes;
+                totalCopiedBytes += copiedBytes;
+
+                // We've run out of bytes in the current block.
+                if (copiedBytes == 0)
+                {
+                    // We hit the end of the blob with the last download call.
+                    if (_position == _length)
+                    {
+                        return totalCopiedBytes;
+                    }
+
+                    // Download the next block
+                    else
+                    {
+                        await Download(async, cancellationToken).ConfigureAwait(false);
+                    }
+                }
+            }
+            while (count > 0);
+            return totalCopiedBytes;
+        }
+
+        private async Task Download(bool async, CancellationToken cancellationToken)
+        {
+            Response<T> response;
+
+            HttpRange range = new HttpRange(_position, _bufferSize);
+
+            try
+            {
+                response = async
+                    ? await _downloadToAsyncFunc(range, _requestConditions, default, cancellationToken).ConfigureAwait(false)
+                    : _downloadToFunc(range, _requestConditions, default, cancellationToken);
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex);
+                throw;
+            }
+
+            Stream networkStream = (Stream)typeof(T).GetProperty("Content").GetValue(response.Value, null);
```

Comment:
Maybe just add another lambda for the client to pass in? If you feel it's too many lambdas, you can look at `PartitionedUploader` for examples on how to organize a lot of lambdas in a class (delegates and constructor taking a struct that has just the delegates).

Reply:
We need to dispose this stream.

Reply:
Add dispose to stream.